src/main/java/org/opentripplanner/middleware/utils/ItineraryUtils.java

```diff
@@ -93,4 +99,139 @@ public static String toQueryString(Map<String, String> params) {
+            !equalsOrPreviousWasNull(previousItineraryLeg.mode, newItineraryLeg.mode) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(previousItineraryLeg.agencyName, newItineraryLeg.agencyName) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.routeLongName,
+                    newItineraryLeg.routeLongName
+                ) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.routeShortName,
+                    newItineraryLeg.routeShortName
+                ) ||
+                !equalsIgnoreCaseOrPreviousWasEmpty(
+                    previousItineraryLeg.headsign,
+                    newItineraryLeg.headsign
+                ) ||
+                !equalsOrPreviousWasNull(
+                    previousItineraryLeg.interlineWithPreviousLeg,
+                    newItineraryLeg.interlineWithPreviousLeg
+                )
+        ) {
+            return false;
+        }
+
+        // make sure the transit trips are scheduled for the same time of the day
+        if (
+            !timeOfDayMatches(
+                previousItineraryLeg.getScheduledStartTime(),
+                newItineraryLeg.getScheduledStartTime()
+            ) || !timeOfDayMatches(
+                previousItineraryLeg.getScheduledEndTime(),
+                newItineraryLeg.getScheduledEndTime()
+            )
+        ) {
+            return false;
+        }
+
+        // if this point is reached, the legs are assumed to match
+        return true;
+    }
+
+    /**
+     * Checks whether two stops (OTP Places) match for the purposes of matching itineraries
+     */
+    private static boolean stopsMatch(Place stopA, Place stopB) {
+        // stop names must match
+        if (!stopA.name.equalsIgnoreCase(stopB.name)) return false;
+
+        // stop code must match
+        if (!equalsIgnoreCaseOrPreviousWasEmpty(stopA.stopCode, stopB.stopCode)) return false;
+
+        // stop positions must be no further than 5 meters apart
+        if (
+            DistanceUtils.radians2Dist(
+                DistanceUtils.distHaversineRAD(stopA.lat, stopA.lon, stopB.lat, stopB.lon),
+                DistanceUtils.EARTH_MEAN_RADIUS_KM
+            ) * 1000 > 5
+        ) {
+            return false;
+        };
+
+        // if this point is reached, the stops are assumed to match
+        return true;
+    }
+
+    /**
+     * Returns true if previous is null. Otherwise, returns Objects.equals.
+     */
+    private static boolean equalsOrPreviousWasNull (Object previous, Object newer) {
+        return previous == null || Objects.equals(previous, newer);
+    }
+
+    /**
+     * Returns true if the previous string was empty. Otherwise, returns if the strings are equal ignoring case.
+     */
+    private static boolean equalsIgnoreCaseOrPreviousWasEmpty(String previous, String newer) {
+        return StringUtils.isEmpty(previous) || previous.equalsIgnoreCase(newer);
```

Comment:
`StringUtils.isEmpty(previous)` can be replaced with `previous.isEmpty()`

Reply:
Do we need to cover the case where `previous` is null too?

Reply:
We definitely need to check whether `previous` is null, so I think using StringUtils is appropriate here.

Reply:
I think the javadoc could be updated to reflect that we're also checking previous for a `null` value.

Reply:
Fixed in ffc3b49b3ad576d936b04758d572efdf3f217ab2